src/js/structs/DSLASTNodes.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * Base class for AST nodes in DSL, representing a token in the user input
+ */
+class ASTNode {
+  constructor(start, end, children=[]) {
+
+    /**
+     * The position array contains an array of tuples, containing the start-end
+     * position of each relevant text portions that composes this token.
+     *
+     * In 99.9% of the cases, this is an array with only 1 tuple. However, when
+     * the `attrib:value1,value2` shorthand is used, this array will contain
+     * 2 tuples, pointing to the `attrib:` and `valueX` parts separately.
+     *
+     * @var {Array}
+     */
+    this.position = [[start, end]];
+
+    /**
+     * Child tokens in this token tree
+     *
+     * @var {Array}
+     */
+    this.children = children;
+
+  }
+}
+
+/**
+ * Combiner (operator) AST node
+ */
+class CombinerNode extends ASTNode {
+  constructor(combinerType, child1, child2) {
+    super(
+      child1.position[0][0],
```

Comment:
Yes, `position` must always contain at least 1 item, with the token position, that being an array of `[start, end]` format. The `ASTNode` constructor guarantees that this is enforced (with the worst case being `[[undefined, undefined]]`.

Reference:
```js
class ASTNode {
  constructor(start, end, children=[]) {
    ...
    this.position = [[start, end]];
    ...
  }
}
```